C Blog example results in "./hello.c:1: error: include file 'stdio.h' not found"

### What is the type of issue?

Example code is not working

### What is the issue?

I'm trying [to follow the first C example](https://bun.sh/blog/compile-and-run-c-in-js) but find this error:

<img width="482" alt="image" src="https://github.com/user-attachments/assets/ade7f342-14fc-4b3c-8416-9d38efa6c222">

Being in MacOS 14.5 (Apple M1) with Bun 1.1.28, I've tried the "basic" things to no avail:
- `xcode-select --install` -> "Command line tools are already installed."
- Double-check files are exactly the same as the first example.
- Changing (temporarily) `void hello() {` to `int main() {` allows me to compile and run the C program as usual:

<img width="265" alt="image" src="https://github.com/user-attachments/assets/e9e1caa1-261e-4165-b114-305e50b3e8f7">

I thought it'd work just like that from reading @Jarred-Sumner's Tweets:
- "No complicated build system. No runtime api."
- "Bun has embedded TinyCC since we introduced FFI support. That means we didn't add any new dependencies to add C compilation support."

So I thought it'd work by default just doing that. Am I missing something perhaps? Do I need to run `bun` in some special way? I don't see any extra instructions in the blog post like "add `--experimental` flag" or anything like that.


### Where did you find it?

First example of https://bun.sh/blog/compile-and-run-c-in-js

Here are my files (they are copy/pasted from the example):

<img width="1044" alt="image" src="https://github.com/user-attachments/assets/12835f5d-13a9-4891-bba7-1a0e073a8996">

Edit: changing the file from `hello.js` to `hello.ts` also didn't help.

Same behavior on Windows 11 (x64):

```
PS C:\Users\xxx\Desktop\bunc> bun .\main.js
1 | import { cc } from "bun:ffi";
2 |
3 | export const {
4 |   symbols: { myRandom },
5 | } = cc({
        ^
error: 1 errors while compiling ./myRandom.c

./myRandom.c:1: error: include file 'stdio.h' not found
      at cc (bun:ffi:87:12)
      at C:\Users\xxx\Desktop\bunc\main.js:5:5

Bun v1.1.28 (Windows x64)
```

and under WSL Ubuntu 24 I get slightly different error (but assume all same root issue of not having proper compiler tooling installed?)
```
error: 1 errors while compiling ./hello.c
In file included from ./hello.c:4:
/usr/include/stdio.h:28: error: include file 'bits/libc-header-start.h' not found
```

> ```
> error: 1 errors while compiling ./hello.c
> In file included from ./hello.c:4:
> /usr/include/stdio.h:28: error: include file 'bits/libc-header-start.h' not found
> ```

@infrahead you can work around this issue by adding this include path, but even after this (on Ubuntu 22.04) i am getting the "library c not found" issue that @stefanos82 is seeing. i have libtcc-dev installed on here so maybe that is the issue. will post here if i find a solution.

```JavaScript
  include: [
    '/usr/include/x86_64-linux-gnu'
  ],
```

I have removed the header file and used forward declaration, plus `main()` function and now I get the following message:

```
hello.ts:1: error: type defaults to int

shell returned 1
```

Can you people double-check it as I'm ready to get some sleep? I'm super tired right now :face_exhaling: 

```c
extern int printf(const char *, ...);

void hello(void)
{
    printf("You can now compile and run C code inside Bun!\n");
}

int main(void)
{
    hello();
    return 0;
}
```


```ts
import { cc } from "bun:ffi";

export const {
    symbols: { hello }
} = cc({
    source: "./hello.c",
    symbols: {
        hello: {
            returns: "void",
            args: [],
        },
    },
});

hello();
```

--- a/src/bun.js/api/ffi.zig
+++ b/src/bun.js/api/ffi.zig
@@ -271,19 +271,19 @@ pub const FFI = struct {
                 }
             } else if (Environment.isLinux) {
                 if (Environment.isX64) {
-                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/x86_64-linux-gnu")) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/x86_64-linux-gnu").isTrue()) {
                         cached_default_system_include_dir = "/usr/include/x86_64-linux-gnu";
                     }
 
-                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/x86_64-linux-gnu")) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/x86_64-linux-gnu").isTrue()) {
                         cached_default_system_library_dir = "/usr/lib/x86_64-linux-gnu";
                     }
                 } else if (Environment.isAarch64) {
-                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/aarch64-linux-gnu")) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/aarch64-linux-gnu").isTrue()) {
                         cached_default_system_include_dir = "/usr/include/aarch64-linux-gnu";
                     }
 
-                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/aarch64-linux-gnu")) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/aarch64-linux-gnu").isTrue()) {
                         cached_default_system_library_dir = "/usr/lib/aarch64-linux-gnu";
                     }
                 }
@@ -352,26 +352,16 @@ pub const FFI = struct {
                 }
 
                 if (Environment.isAarch64) {
-                    switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/opt/homebrew/include")) {
-                        .result => |exists| {
-                            if (exists) {
-                                if (TCC.tcc_add_include_path(state, "/opt/homebrew/include") == -1) {
-                                    debug("TinyCC failed to add library path", .{});
-                                }
-                            }
-                        },
-                        .err => {},
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/opt/homebrew/include").isTrue()) {
+                        if (TCC.tcc_add_include_path(state, "/opt/homebrew/include") == -1) {
+                            debug("TinyCC failed to add library path", .{});
+                        }
                     }
 
-                    switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/opt/homebrew/lib")) {
-                        .result => |exists| {
-                            if (exists) {
-                                if (TCC.tcc_add_library_path(state, "/opt/homebrew/lib") == -1) {
-                                    debug("TinyCC failed to add library path", .{});
-                                }
-                            }
-                        },
-                        .err => {},
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/opt/homebrew/lib").isTrue()) {
+                        if (TCC.tcc_add_library_path(state, "/opt/homebrew/lib") == -1) {
+                            debug("TinyCC failed to add library path", .{});
+                        }
                     }
                 }
             } else if (Environment.isLinux) {
@@ -389,48 +379,16 @@ pub const FFI = struct {
             }
 
             if (Environment.isPosix) {
-                switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/local/include")) {
-                    .result => |exists| {
-                        if (exists) {
-                            if (TCC.tcc_add_sysinclude_path(state, "/usr/local/include") == -1) {
-                                debug("TinyCC failed to add library path", .{});
-                            }
-                        }
-                    },
-                    .err => {},
+                if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/local/include").isTrue()) {
+                    if (TCC.tcc_add_sysinclude_path(state, "/usr/local/include") == -1) {
+                        debug("TinyCC failed to add library path", .{});
+                    }
                 }
 
-                switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include")) {
-                    .result => |exists| {
-                        if (exists) {
-                            if (TCC.tcc_add_sysinclude_path(state, "/usr/include") == -1) {
-                                debug("TinyCC failed to add library path", .{});
-                            }
-                        }
-                    },
-                    .err => {},
-                }
-
-                switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/local/lib")) {
-                    .result => |exists| {
-                        if (exists) {
-                            if (TCC.tcc_add_library_path(state, "/usr/local/lib") == -1) {
-                                debug("TinyCC failed to add library path", .{});
-                            }
-                        }
-                    },
-                    .err => {},
-                }
-
-                switch (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib")) {
-                    .result => |exists| {
-                        if (exists) {
-                            if (TCC.tcc_add_library_path(state, "/usr/lib") == -1) {
-                                debug("TinyCC failed to add library path", .{});
-                            }
-                        }
-                    },
-                    .err => {},
+                if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/local/lib").isTrue()) {
+                    if (TCC.tcc_add_library_path(state, "/usr/local/lib") == -1) {
+                        debug("TinyCC failed to add library path", .{});
+                    }
                 }
             }
 

--- a/src/bun.js/api/ffi.zig
+++ b/src/bun.js/api/ffi.zig
@@ -241,6 +241,67 @@ pub const FFI = struct {
 
         pub const default_tcc_options: [:0]const u8 = "-std=c11 -Wl,--export-all-symbols -g -O2";
 
+        var cached_default_system_include_dir: [:0]const u8 = "";
+        var cached_default_system_library_dir: [:0]const u8 = "";
+        var cached_default_system_include_dir_once = std.once(getSystemRootDirOnce);
+        fn getSystemRootDirOnce() void {
+            if (Environment.isMac) {
+                var which_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+
+                var process = bun.spawnSync(&.{
+                    .stdout = .buffer,
+                    .stdin = .ignore,
+                    .stderr = .ignore,
+                    .argv = &.{
+                        bun.which(&which_buf, bun.sliceTo(std.c.getenv("PATH") orelse "", 0), Fs.FileSystem.instance.top_level_dir, "xcrun") orelse "/usr/bin/xcrun",
+                        "-sdk",
+                        "macosx",
+                        "-show-sdk-path",
+                    },
+                    .envp = std.c.environ,
+                }) catch return;
+                if (process == .result) {
+                    defer process.result.deinit();
+                    if (process.result.isOK()) {
+                        const stdout = process.result.stdout.items;
+                        if (stdout.len > 0) {
+                            cached_default_system_include_dir = bun.default_allocator.dupeZ(u8, strings.trim(stdout, "\n\r")) catch return;
+                        }
+                    }
+                }
+            } else if (Environment.isLinux) {
+                if (Environment.isX64) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/x86_64-linux-gnu")) {
+                        cached_default_system_include_dir = "/usr/include/x86_64-linux-gnu";
+                    }
+
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/x86_64-linux-gnu")) {
+                        cached_default_system_library_dir = "/usr/lib/x86_64-linux-gnu";
+                    }
+                } else if (Environment.isAarch64) {
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/include/aarch64-linux-gnu")) {
+                        cached_default_system_include_dir = "/usr/include/aarch64-linux-gnu";
+                    }
+
+                    if (bun.sys.directoryExistsAt(std.fs.cwd(), "/usr/lib/aarch64-linux-gnu")) {
+                        cached_default_system_library_dir = "/usr/lib/aarch64-linux-gnu";
+                    }
+                }
+            }
+        }
+
+        fn getSystemIncludeDir() ?[:0]const u8 {
+            cached_default_system_include_dir_once.call();
+            if (cached_default_system_include_dir.len == 0) return null;
+            return cached_default_system_include_dir;
+        }
+
+        fn getSystemLibraryDir() ?[:0]const u8 {
+            cached_default_system_include_dir_once.call();
+            if (cached_default_system_library_dir.len == 0) return null;
+            return cached_default_system_library_dir;
+        }
+
         pub fn compile(this: *CompileC, globalThis: *JSGlobalObject) !struct { *TCC.TCCState, []u8 } {
             const state = TCC.tcc_new() orelse {
                 globalThis.throw("TinyCC failed to initialize", .{});
@@ -266,17 +327,27 @@ pub const FFI = struct {
             }
 
             if (Environment.isMac) {
-                if (bun.getenvZ("SDKROOT")) |sdkroot| {
-                    const include_dir = bun.path.joinAbsStringBufZ(sdkroot, &pathbuf, &.{ "usr", "include" }, .auto);
-                    if (TCC.tcc_add_sysinclude_path(state, include_dir.ptr) == -1) {
-                        globalThis.throw("TinyCC failed to add sysinclude path", .{});
-                        return error.JSException;
-                    }
+                add_system_include_dir: {
+                    const dirs_to_try = [_][]const u8{
+                        bun.getenvZ("SDKROOT") orelse "",
+                        getSystemIncludeDir() orelse "",
+                    };
 
-                    const lib_dir = bun.path.joinAbsStringBufZ(sdkroot, &pathbuf, &.{ "usr", "lib" }, .auto);
-                    if (TCC.tcc_add_library_path(state, lib_dir.ptr) == -1) {
-                        globalThis.throw("TinyCC failed to add library path", .{});
-                        return error.JSException;
+                    for (dirs_to_try) |sdkroot| {
+                        if (sdkroot.len > 0) {
+                            const include_dir = bun.path.joinAbsStringBufZ(sdkroot, &pathbuf, &.{ "usr", "include" }, .auto);
+                            if (TCC.tcc_add_sysinclude_path(state, include_dir.ptr) == -1) {
+                                globalThis.throw("TinyCC failed to add sysinclude path", .{});
+                                return error.JSException;
+                            }
+
+                            const lib_dir = bun.path.joinAbsStringBufZ(sdkroot, &pathbuf, &.{ "usr", "lib" }, .auto);
+                            if (TCC.tcc_add_library_path(state, lib_dir.ptr) == -1) {
+                                globalThis.throw("TinyCC failed to add library path", .{});
+                                return error.JSException;
+                            }
+                            break :add_system_include_dir;
+                        }
                     }
                 }
 
@@ -301,6 +372,18 @@ pub const FFI = struct {
                             }
                         },
                         .err => {},
+                    }
+                }
+            } else if (Environment.isLinux) {
+                if (getSystemIncludeDir()) |include_dir| {
+                    if (TCC.tcc_add_sysinclude_path(state, include_dir) == -1) {
+                        debug("TinyCC failed to add library path", .{});
+                    }
+                }
+
+                if (getSystemLibraryDir()) |library_dir| {
+                    if (TCC.tcc_add_library_path(state, library_dir) == -1) {
+                        debug("TinyCC failed to add library path", .{});
                     }
                 }
             }

--- a/src/bun.js/node/types.zig
+++ b/src/bun.js/node/types.zig
@@ -99,6 +99,14 @@ pub fn Maybe(comptime ReturnTypeT: type, comptime ErrorTypeT: type) type {
             return .{ .err = ErrorType{} };
         }
 
+        pub fn isTrue(this: @This()) bool {
+            if (comptime ReturnType != bool) @compileError("This function can only be called on bool");
+            return switch (this) {
+                .result => |r| r,
+                else => false,
+            };
+        }
+
         pub fn unwrap(this: @This()) !ReturnType {
             return switch (this) {
                 .result => |r| r,

